### DISCUSSION TITLE
MySQL 에선 VARCHAR를 쓸 때 시간적인 성능이 CHAR보다 안좋은 이유?

### DISCUSSION THREAD

**Discussion by jschoi-96** [Hanjaemo/DB-Study#8] [Q&A]
TODO

**Comment by Hanjaemo** [comment]
MySQL에서 CHAR 타입은 '고정 크기 문자열'을 의미하며, 저장될 문자열의 길이가 최대 길이보다 작으면 나머지를 space로 채워서 저장합니다.

반면에 VARCHAR 타입은 '가변 크기 문자열'을 의미하며, 저장될 문자열의 길이가 최대 길이보다 작더라도 해당 문자열만 저장합니다.
즉, VARCHAR 타입은 데이터를 읽어올 때 해당 문자열의 길이가 얼마인지 계산하는 연산을 거치게 되고, 이로 인해 CHAR 타입에 비해 검색 속도 및 읽기 속도가 더 느립니다.

하지만 이러한 **속도 차이는 실제로 체감하지 못할 정도로 작기 때문에**, VARCHAR의 시간적인 성능에 대한 의구심을 가진다는 것은 이론 중심적인 생각이라고 합니다.

따라서 시간적인 성능 때문에 CHAR 타입 사용을 고민하기 보다는, VARCHAR 타입을 통해 관리의 편의성과 프로젝트의 유연성을 확보하는 것이 더 좋다고 생각합니다.
하지만 이 또한 제 개인적인 생각으로, 팀 상황에 따라 적절한 선택지를 고르는 것이 가장 중요할 것 같습니다.

**Comment by mithzinf** [comment]
CHAR와 VARCHAR 둘 다 문자열을 저장하는 데이터 타입이지만 엄밀히 따지면, CHAR는 고정 길이의 문자열을 저장하고, VARCHAR는 가변 길이의 문자열을 저장하는 데에서 오는 차이가 곧 시간적 성능을 가르는 차이점이라고 생각됩니다. 

- CHAR : 고정 길이 문자열을 저장, 지정된 길이에 미치지 못하는 경우 -> 공백으로 채워짐 ∴데이터 검색 시 시간적 성능 우위
- VARCHAR : 가변 길이 문자열을 저장, 실제 데이터 길이에 따라 공간을 달리 사용하므로 공간 효율성 우위, but 각각의 길이를 가진 데이터 정보를 저장해야 하므로 추가적인 입출력 과정이 필요할 수도 있음

위의 두 가지 문자열 타입의 특징이 시간적 성능에 미치는 영향과 연관이 되어 있습니다

IF CHAR 사용할 시, 길이가 고정되어 있는 터라 문자열 조작이 덜 필요하며 인덱스 조회가 빠릅니다. (시간적 성능 우위), 특히 데이터 크기가 일정하고 자주 변경이 되지 않는 경우에 적합함

IF VARCHAR 사용할 시, 가변 길이 데이터를 저장하게 되므로 그에 따른 추가적인 정보도 관리를 해야하기 떄문에, CHAR에 비해 시간적 성능 면에서는 아쉬운 점이 있음
특히, 데이터 길이가 자주 변하거나, 다양한 길이의 데이터인 경우에 성능 저하가 발생할 수 있음


결론 : 무조건적으로 특정 데이터 타입만이 명답이라는 것은 아니고, DB 설계시, 데이터의 성격과 어떻게 사용하게 될 것인지를 고려하여 적절한 타입을 선택하는 것이 중요하다고 생각합니다.
만약에, 데이터의 길이가 다 비슷비슷하다면 CHAR 데이터 타입을 쓰는 것이 더 맞겠고, 가변 길이 데이터에는 VARCHAR를 쓰는 것이 더 옳을 것 같네요!
